src/main/java/emissary/core/BaseDataObject.java

```diff
@@ -221,6 +223,21 @@ protected DataState getDataState() {
         }
     }
 
+    @Override
+    public void checkAndResetArrayHashMap(final String placeName) {
+        for (Map.Entry<byte[], String> e : arrayHashMap.entrySet()) {
+            if (!ByteUtil.sha256Bytes(e.getKey()).equals(e.getValue())) {
```

Comment:
I think this can only ever be true if the DataState is BYTE_ARRAY_ONLY (i.e., `theData` != null), because that's the only scenario for which the byte[] stored in the Map key will have any live references. In the CHANNEL_ONLY scenario, the Map entries are essentially immutable so the hash of the key will ALWAYS equal it's cached has value.

Reply:
Ouch.  This isn't true;  in the CHANNEL_ONLY scenario, every call to `bdo.data()` will create a new entry in the map, with the key being the returned byte[]. The byte[] object references will differ even it the bytes in the array are identical, so we'll store multiple copies of the payload bytes in the Map if `bdo.data()` is called multiple times within the same place.